### PR TITLE
feat(ST-150): update user password after otp verification

### PR DESCRIPTION
**Pull request by faysal-alam-sazim**
## Description of Change

- Added PATCH `users/update-password` API endpoint
- Added necessary service method
- Checked isOtpVerified or not before updating password

## Associated Ticket Link(s)

- https://trello.com/c/z6ukXYEk

## Related Pull Request(s)

- N/A

### CODE REVIEW

File: src/users/users.service.ts [hunk 1]
```diff
@@ -61,4 +62,35 @@ export class UsersService {
...
   }
+
+  isOtpVerified(user: User) {
+    const latestVerifyUser = user.verifyUser
+      .getItems()
+      .sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime())[0];
+
+    if (!latestVerifyUser) {
+      throw new Error("No verification record found for this user.");
+    }
```
Comment (ishmam-iqbal-sazim) [inline src/users/users.service.ts:73]: Do you need to throw an error here? Boolean return should be fine

File: src/users/users.service.ts [hunk 2]
```diff
@@ -61,4 +62,35 @@ export class UsersService {
...
+      if (this.isOtpVerified(user)) {
+        user.password = await this.hashPassword(newPassword);
+
+        await this.usersRepository.getEntityManager().persistAndFlush(user);
+      } else {
+        throw new UnauthorizedException("OTP isn't verified");
+      }
+    } catch (error) {
+      throw new BadRequestException(error);
+    }
```
Comment (ishmam-iqbal-sazim) [inline src/users/users.service.ts:94]: Do you need this try catch?

File: src/users/users.service.ts [hunk 3]
```diff
@@ -61,4 +62,35 @@ export class UsersService {
...
+        { populate: ["verifyUser"] },
+      );
+
+      if (this.isOtpVerified(user)) {
+        user.password = await this.hashPassword(newPassword);
+
+        await this.usersRepository.getEntityManager().persistAndFlush(user);
+      } else {
+        throw new UnauthorizedException("OTP isn't verified");
+      }
```
Comment (ishmam-iqbal-sazim) [inline src/users/users.service.ts:91]:
```suggestion
      if (!this.isOtpVerified(user)) {
        throw new UnauthorizedException("OTP isn't verified");
      }
      
      user.password = await this.hashPassword(newPassword);

      await this.usersRepository.getEntityManager().persistAndFlush(user);
```

Comment (ishmam-iqbal-sazim) [reply src/users/users.service.ts:91]: Also, Exception should be Forbidden (403)

**Review by ragib-hossain-sazim — APPROVED** [review]

File: src/users/users.service.ts [hunk 4]
```diff
@@ -71,4 +72,28 @@ export class UsersService {
...
+
+    const user = await this.usersRepository.findOneOrFail({ email }, { populate: ["verifyUser"] });
+
+    if (!this.isOtpVerified(user)) {
+      throw new UnauthorizedException("OTP isn't verified");
+    }
+    user.password = await this.hashPassword(newPassword);
+
+    await this.usersRepository.getEntityManager().persistAndFlush(user);
+  }
```
Comment (ishmam-iqbal-sazim) [inline src/users/users.service.ts:98]:
```
  async updateUserPassword(updatePasswordDto: UpdatePasswordDto) {
    const { email, newPassword, otp } = updatePasswordDto;

    const user = await this.usersRepository.findOneOrFail({ email }, { populate: ["verifyUser"] });
    
    const verifyUser = await this.verifyUsersRepository.findOneOrFail({ user: user, otp: otp })
    
    const isOtpValid = verifyUser.isChecked && verifyUser.expiryDate > new Date();
    
    if (!isOtpValid) {
        throw new UnauthorizedException("OTP isn't verified or has expired");
    }
    user.password = await this.hashPassword(newPassword);

    await this.usersRepository.getEntityManager().persistAndFlush(user);
  }
```
[NIT] - I would do something like this. I think this implementation is more intuitive and secure. But

**Review by ishmam-iqbal-sazim — APPROVED** [review]
Left a comment

**Pull request merged**